[codechandler] Check and correct stream codec info

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
If manifest provide malformed info like wrong codec name side effects like broken audio/video can happen on playback,
so we have to correct these info by using our demuxers/readers

the only side effect is since this is performed just before the playback the kodi streams still have wrong info,
not so bad but for example if we choose the stream quality manually on kodi GUI we will read the wrong codec name but playback will works correctly

I have add code to fix codec name on all readers/codec handlers ~with exception of `ADTSReader`
I was not able to understand why i could not read the `AdtsType` by using `ADTSReader::GetInformation`~ (`ADTSReader` included)

i have also moved all related code for audio cases in a generic audio codec handler thats its more clear
this include an additional bug fix on the use of `AP4_GenericAudioSampleDescription` cast (previously in the CodecHandler.cpp) was always failing because the parent class is `AP4_AudioSampleDescription`

i have also reworked the workaround `GetAudioCodec` in order to be easier to maintain
https://github.com/xbmc/inputstream.adaptive/blob/21.1.3-Omega/src/parser/HLSTree.cpp#L104-L131

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
try to fix #1254
## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
hacked a stream manifest to try simulate the problem of issue
tested couple of webm and a dash sample

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

CastagnaIT committed:
[HLSTree] Add m4s ext case to detect container type

CastagnaIT committed:
[codechandler] Add audio codec handler

CastagnaIT committed:
[TSReader] Cleanup GetInformation code

CastagnaIT committed:
[TSReader] Check if codec name is changed

CastagnaIT committed:
[WebmReader] Check stream codec name info

CastagnaIT committed:
[ADTSReader] Check stream codec name info

CastagnaIT committed:
[StringUtils] Add SplitToVec method

CastagnaIT commented:
@matthuisman have you time to do a fast test with Disn€y+ with test build of this PR?

i need a simply thing to check, you have to try play some videos that contains both audio track types, dolby+ audio and AAC tracks, then check if codec names shown on kodi GUI audio track list are correct and not wrogly mixed (e.g. if you select AAC track the sound should be AAC and not dolby+)

CastagnaIT committed:
[HLSTree] Improved GetAudioCodec easier to maintain

CastagnaIT committed:
[HLSTree] Add codec and channels to adpsets of groups

This help AdaptiveTree::SortTree() on adaptation sets merging
otherwise easily cause false positive with CAdaptationSet::IsMergeable
that will result for example in a lack of audio tracks when in playback

CastagnaIT committed:
[CRepresentation] Cleanup codecs string

matthuisman commented:
sure. trying now

matthuisman commented:
something doesnt seem right
Current released IA
![screenshot00005](https://github.com/xbmc/inputstream.adaptive/assets/6225961/64f6d6ff-ed5a-4726-87be-a4de2b4be765)


With this PR:
![screenshot00000](https://github.com/xbmc/inputstream.adaptive/assets/6225961/5040b4a6-c0ff-4967-95bd-17e0cc5ad90a)
![screenshot00000](https://github.com/xbmc/inputstream.adaptive/assets/6225961/07638264-63fa-4224-9168-9579eb432cd0)
![screenshot00002](https://github.com/xbmc/inputstream.adaptive/assets/6225961/f3925419-51f9-4d99-a9a9-65d870301753)
![screenshot00004](https://github.com/xbmc/inputstream.adaptive/assets/6225961/cbd5269c-193f-494c-a088-064789fc56d3)

CastagnaIT commented:
can you provide the master manifest of screenshot?

matthuisman commented:
actually. after disabling proxy (i had commented some stuff that caused off stuff) and trying again - they appear the same

BEFORE:
![screenshot00003](https://github.com/xbmc/inputstream.adaptive/assets/6225961/c8d0c3df-883d-4208-8c24-6244278ea6e0)
AFTER:
![screenshot00004](https://github.com/xbmc/inputstream.adaptive/assets/6225961/cfe16289-3b63-4387-b4a5-722dbd534e05)


But, when I play the DD+ 6 channel - it changes to 2 channel after few seconds
audio stream still shows the 6 channels though
Same with the 16 channels (changes to 2 but still 16)
The 2 channel streams seem ok (stay on 2)

Attached is the manif
[manifest_1689924724_master.txt](https://github.com/xbmc/inputstream.adaptive/files/12117579/manifest_1689924724_master.txt)
est

Ronny-nerd commented:
Hello matthuisman ...have you for example tried to start an episode from these series via an Android device...what then happens with the sound?

Secret Invasion
The Mandalorian
Obi-Wan Kenobi 
The Book of Boba Fett
Star Wars: Andor
Star Wars: Tales of the Jedi

Ronny-nerd commented:
Hardware: Nvidia Shield TV 2015
No AVR and no soundbar

Kodi 21 Omega Alpha 2
With InputStream Adaptive 21.1.0 and Kodi 20.2 Nexus there is no such problem.

For me, there is no sound at the start of an episode in said series, only when you manually switch the sound.

[Problem 3 InputStream Adaptive.zip](https://github.com/xbmc/inputstream.adaptive/files/12122588/Problem.3.InputStream.Adaptive.zip)

![072123115009](https://github.com/xbmc/inputstream.adaptive/assets/65504837/200eda3a-6e1c-4614-81a6-250842e97064)
![072123115308](https://github.com/xbmc/inputstream.adaptive/assets/65504837/0864e790-7ea5-48ff-a88f-47b264d9c24b)

CastagnaIT commented:
@Ronny-nerd if you have some problem open an issue dont use development threads

Ronny-nerd commented:
Ok CastagnaIT ...have opened a new one.

https://github.com/xbmc/inputstream.adaptive/issues/1335

CastagnaIT committed:
[AudioCodecHandler] Old sample description versions support max 2ch

CastagnaIT commented:
@matthuisman found the problem now should be all good

Ronny-nerd commented:
> Strictly speaking. After disabling the proxy (I had commented on some things that caused errors) and tried again, they look the same

Where did you disable the proxy?

matthuisman commented:
tested with latest change - works good :)

glennguy commented on src/parser/HLSTree.cpp:
```diff
@@ -101,35 +101,21 @@ ContainerType DetectContainerTypeFromExt(std::string_view extension)
-    return CODEC::FOURCC_MP4A;
+  // Some manifests can provide CODECS attribute with more audio codecs
+  // it seems that usually the first in order of apparence is the one referring to the specified
+  // AUDIO group attribute, so we return the first one found
```
Great find here. To me this was always a problem that would be difficult to solve but if they're ordered then our lives are much easier :)

CastagnaIT replied:
this can works for case with group attribute,
but im quite sure that this is still wrong for the case when a video TS have embedded/included multiple audio streams (without audio group)
also im not sure that currently the parser is able to handle TS stream with more included audio streams
it seems to me that it is not even considered, more likely if/when we will find a sample stream to make tests will require many more code adjustments

glennguy replied:
Yes TS case is a different baby. Same goes for fMP4 with multiple tracks (audio/video, audio/audio or video/audio/audio)... we can add to the long list of wishes! :)

glennguy commented on src/codechandler/HEVCCodecHandler.cpp:
```diff
@@ -69,26 +72,50 @@ bool HEVCCodecHandler::ExtraDataToAnnexB()
+  if (fourcc > 0 && info.GetCodecFourCC() != fourcc)
+  {
+    info.SetCodecFourCC(fourcc);
+    isChanged |= true;
```
is the `|=` operator redundant when assigning to true?

CastagnaIT replied:
right! done

glennguy reviewed (approved):
Looks great, just one comment which I'll let you decide on

CastagnaIT committed:
[codechandler] Check stream codec name info

CastagnaIT committed:
[HLSTree] Fallback to TS container also for audio

We have recently add in OpenStream() fallback from TS to ADTS
when initialization fails, so also for AUDIO we can set TS
